### ISSUE TITLE
Feature request: framed text boxes

### ISSUE THREAD

**Issue by TobiasZawada** [misohena/el-easydraw#32]
It would be great to have framed text boxes. This would especially be a nice feature for drawing UML class diagrams.
Several variants come to my mind:

1. The frame could be an additional option for the text tool. This should also include the form of the frame, e.g., circle, rectangle, rectangle with rounded corners, and so on.
2. The text feature could be added to other tools, like the rectangle tool or the circle tool.
3. Texts in groups could be kept easily editable. In this case the group should adapt its scale or its extent to the contained text.

**Comment by misohena** [comment]
(Translated by Chat-GPT)
This is a feature I’ve wanted since the early stages of development, but there’s still no prospect of its implementation. There are some reasons for this, but explaining them here in detail (and in proper English) is too much trouble, so I won’t.

It might naturally be realized someday as improvements progress.

Given how difficult this is, at the very least, when something is demanded by a complete stranger with no return, it makes me lose the desire to do it, even if I wanted to do it before.

Since this negatively affects my motivation, I will close this issue.

(Original Message)
これは私も開発当初から望んでいる機能ですが未だに実現の見通しすら立っていません。実現出来ていないのにはそれなりの理由がありますが、ここで詳しく(英語に正しく翻訳できる形式で)説明するのは面倒くさすぎるのでしません。

改良を進めるうちにそのうち自然と実現される日が来るかもしれません。

これだけ大変なことを少なくとも赤の他人に何の見返りも無しに要求されたらやりたかったものもやりたくなくなってしまいます。

私のやる気に悪影響があるのでIssueをクローズします。

**Comment by TobiasZawada** [comment]
Yes, I agree that this is not as easy as the arrow heads. One has to adapt the box to the text or visa versa.

The gluing of the text to the box is a rough workaround working under sever conditions.
One severity is related to the fact that one cannot use formatting for pieces of text. So one has to patch together several text boxes when one wants parts of the text formatted differently (for putting emphasis on some part of the text). But this does not work with gluing.

Would it be possible to html-format parts of the text of one text box? If you want, I could open a new feature request for this.

I think the easiest quick-and-dirty-solution would be to optionally avoid transforming < into &langle; and by keeping the html-subtree for the text instead of using just the text node.

**Comment by misohena** [comment]
(Translated by Google Translate)
Does the html-format parts mean embedding HTML in SVG with foreignObject? I think librsvg probably doesn't support it, but I'm not sure (unconfirmed).
# During development, I'm sometimes surprised by the differences between web browsers and librsvg.

(Original Message)
html-format partsというのはforeignObjectでSVGの中にHTMLを埋め込むということでしょうか？　おそらくlibrsvgは対応していないのではないかと思いますがどうでしょう(未確認)。
# 開発中、Webブラウザとlibrsvgとの差に驚くことがたまにあります。

**Comment by TobiasZawada** [comment]
Yes, you are right. HTML is the wrong thing. The right thing is the `<tspan>` tag of SVG.

Example:
```
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="560" height="420" version="1.1">
  <g id="edraw-body">
    <rect x="0" y="0" width="560" height="420" id="edraw-background" stroke="none" fill="#fff" />
    <text y="160" x="260" fill="#222" text-anchor="middle" font-size="18" font-family="sans-serif">Test of <tspan style="font-weight:bold">html tags for formatting</tspan>.</text>
  </g>
</svg>
```
GUI support for the formatting would be great. But, the right handling should be thoroughly considered.

Maybe, raw input of the svg xml content of the `<text>` object would be an optional expert input that could be kept in the final version.

**Comment by misohena** [comment]
(Translated by Chat-GPT)
When writing UML diagrams, italics or underlines are something you'd probably want.

I've added a feature to specify the glue position. It's a bit cumbersome since it requires numerical input, but it might still be of some use.

(Original)
UMLを書くなら斜体や下線は欲しいですよね。

接着機能の接着位置を指定出来るようにしてみました。数値入力なので使いづらいですが何かの足しになるかもしれません。

![20250104-glue-position](https://github.com/user-attachments/assets/03a7133f-f6b5-4550-9e93-627987af82d1)

**Comment by misohena** [comment]
Partial bold, italics and underlining are now possible with commit 54feb9e5982a3f38e561cdd28494bed23784bedb.

コミット 54feb9e5982a3f38e561cdd28494bed23784bedb によって部分的な太線、斜体、下線はできるようになりました。

**Comment by TobiasZawada** [comment]
I pulled the last version.
There seems to be a glitch:

` the function ‘edraw-svg-read-propertized-text’ is
    not known to be defined.`

**Comment by misohena** [comment]
(Translated by Chat-GPT)
I feel like I might have mentioned this before, but I prefer not to put unpredictable elements into edraw-body.

Edraw imposes various constraints on its internal structure to enable GUI-based editing. These constraints are what make it convenient to edit in the first place.

You mentioned allowing raw XML editing, but modifying the mechanisms built for GUI editing to accept raw XML as text content is not as simple as it sounds.

If effort is to be spent, it would be better directed toward expanding what can be achieved through proper, standard editing methods.

That said, there is a way to bypass most of those mechanisms and forcibly edit using raw XML.

For example, you can create a command like the following:
```elisp
(require 'edraw)

(defun my-edraw-edit-selected-text-content ()
  (interactive)
  (let* ((editor (edraw-current-editor))
         (text-shape (or (edraw-cast (edraw-last-selected-shape editor)
                                     'edraw-shape-text)
                         (error "No text selected")))
         (old-xml (my-edraw-get-inner-xml text-shape))
         (new-xml (read-string "New XML: " old-xml)))
    (unless (equal new-xml old-xml)
      (my-edraw-set-inner-xml text-shape new-xml old-xml))))

(defun my-edraw-get-inner-xml (shape)
  (mapconcat (lambda (child) (edraw-svg-encode child nil nil))
             (edraw-dom-children (edraw-element shape))))

(defun my-edraw-set-inner-xml (shape new-xml old-xml)
  (let ((g (or (edraw-svg-decode-xml (concat "<g>" new-xml "</g>") nil)
               (error "Invalid XML")))
        (element (edraw-element shape)))
    (edraw-dom-remove-all-children element)
    (edraw-dom-append-children element (edraw-dom-children g))
    (edraw-push-undo (edraw-get-editor shape)
                     'my-edraw-set-inner-xml
                     (list #'my-edraw-set-inner-xml shape old-xml new-xml))
    (edraw-on-shape-changed shape 'my-edraw-set-inner-xml)))
```

By selecting a text object and running M-x my-edraw-edit-selected-text-content, you should be able to directly edit its content.

Please note that librsvg's text-related functionality has many bugs and unimplemented features, so take care when using it.

I came up with a simple piece of code and wrote it down. 😁

(Original Message)
前にも書いたような気もしますが、あまりedraw-bodyの中に予測できないものを入れたくないんですよね。

edrawはGUIで編集できるようにするために、内部構造に様々な制約があります。その制約のおかげで便利に編集が出来るとも言えるわけです。

生のXMLを書けるようにするだけと言いますが、GUI編集のために構築した様々なメカニズムを、テキストの内容として生のXMLを受け付けるように改造するのは言うほど簡単ではありません。

そこに労力を費やすなら、ちゃんとした正規の編集方法でできることを増やす方に労力を費やした方が良いでしょう。

とは言えそういったメカニズムの大半を迂回して無理矢理生のXMLで編集する方法はあります。

例えば次のようなコマンドを作成して……

(CODE)

テキストを選択して M-x my-edraw-edit-selected-text-content を実行すれば内容を直接編集することができると思います。

ちなみにテキスト関連はlibrsvgのバグや未実装が多いのでご注意を。

簡単なコードを思いついたのでちょっと書いてみました。😁

**Comment by TobiasZawada** [comment]
>I feel like I might have mentioned this before, but I prefer not to put unpredictable elements into edraw-body.

Yeah, you already mentioned that when we discussed about general SVG features. I remember.

I have a slight feeling that you are running into a dead end when you couple the formatting of SVG text to text properties of Emacs. Maybe, the edited text object should be SVG. This SVG should not be edited directly in the drawing, since this might slow down editing drastically, It should be a small svg object only holding the text. Then the implementation of all formatting operations that libsvg allows should be possible on this text object. One could implement those features one after the other whenever one has the required spare time.

...Just an idea. I can imagine that this is not as simple as transforming text properties into SVG since this requires the text-edit-features on SVG side. But, maybe, one can mirror the editing in some invisible temporary buffer, put there properties with the `tspan`-Attributes on the relevant stretches of text and transform it back into an SVG text object on-the-fly. A difficulty with this approach would be the synchronization of the cursor position between the SVG text object and the text in the invisible buffer.

As I said, just an idea.

**Comment by TobiasZawada** [comment]
Take your time. Being able to propertize the text is already a big step forward.
When I split the text of one text box beforehand I increased the text size of the first part a little. But, meanwhile I think that setting it bold is enough for emphasizing it sufficiently. So I do not need to split the text anymore.

For translating your text you could give https://www.deepl.com/ a try. It is not as general as ChatGPT but focuses on translation of human languages. It is also AI-based and may deliver results that better meet your requirements.
